File: ChessLab/Analysis/MoveExplanation.swift
import ChessKit

/// Pourquoi un coup était mauvais — en une phrase vraie.
///
/// Le bandeau coach sait dire « e5 — Erreur, −12 % » : il sait donc COMBIEN un
/// coup coûte, et n'a jamais su dire CE QUI le punit. L'utilisateur apprend
/// qu'il a eu tort, pas ce qu'il n'a pas vu, et rejouera le même coup.
///
/// Cette structure porte la réponse, lue sur la **réfutation du moteur** — la
/// variante qu'il enchaîne après le coup joué. Aucun modèle de langage n'entre
/// ici : tout est rejoué sur un plateau, donc rien ne peut être inventé. C'est
/// moins bavard qu'un coach improvisé, et c'est vérifiable.
struct MoveExplanation: Equatable {
    /// Le motif qui punit, quand il est nommable. `nil` est fréquent et normal :
    /// beaucoup de fautes se résument à une perte de matériel sans figure.
    var motif: TacticalMotif?
    /// Matériel net perdu sur la réfutation (en points), du point de vue du
    /// joueur qui vient de jouer. `nil` quand la ligne ne coûte rien de net.
    var materialLoss: Int?
    /// Le premier coup de la réfutation, en SAN anglais (le standard stocké —
    /// la francisation est une affaire d'affichage, voir ``SANFormatter``).
    var refutationSAN: String

    /// La phrase, dans la langue active de l'app.
    ///
    /// Construite à la LECTURE et non à l'analyse : l'explication est mise en
    /// cache pour toute la session, et un changement de langue en cours de
    /// route doit la retraduire — pas la laisser figée dans la langue qui
    /// avait cours au moment du calcul.
    func sentence(notation: PieceNotation) -> String {
        let san = SANFormatter.display(refutationSAN, notation: notation)

        guard let motif else {
            guard let materialLoss else { return "" }
            return LocalizationController.string(
                "%@ : vous perdez %lld points de matériel.", san, materialLoss
            )
        }

        let head: String
        var mentionsCost = true

        switch motif {
        case let .checkmate(inMoves, isBackRank):
            // Le matériel ne veut plus rien dire quand la ligne mate.
            mentionsCost = false
            head = switch (inMoves <= 1, isBackRank) {
            case (true, true): LocalizationController.string("%@ : mat du couloir.", san)
            case (true, false): LocalizationController.string("%@ : et c'est mat.", san)
            case (false, true):
                LocalizationController.string("%@ : mat du couloir en %lld coups.", san, inMoves)
            case (false, false):
                LocalizationController.string("%@ : mat en %lld coups.", san, inMoves)
            }

        case let .hangingPiece(kind, _):
            // Le coût est le nom de la pièce : le répéter en points serait du bruit.
            mentionsCost = false
            head = LocalizationController.string(
                "%@ : votre %@ était en prise.", san, PieceNaming.frenchKindLowercase(kind)
            )

        case let .fork(_, _, targets):
            // Deux cibles nommées suffisent — une fourchette triple se raconte
            // aussi bien par ses deux plus grosses prises.
            let named = targets.prefix(2).map(PieceNaming.frenchKindLowercase)
            head = named.count == 2
                ? LocalizationController.string(
                    "%@ : fourchette sur votre %@ et votre %@.", san, named[0], named[1]
                )
                : LocalizationController.string("%@ : fourchette.", san)

        case .discoveredCheck:
            // La pièce qui démasque n'est PAS nommée : « son fou »/« sa tour »
            // demanderait un accord de genre par pièce, pour un gain nul —
            // la flèche sur le plateau montre déjà laquelle.
            head = LocalizationController.string("%@ : échec à la découverte.", san)

        case let .pin(victim, behind):
            // Tournure NOMINALE (« clouage de votre… ») et non « votre tour est
            // clouée » : le participe s'accorderait en genre avec la pièce, ce
            // qu'un gabarit à trous ne sait pas faire.
            head = LocalizationController.string(
                "%@ : clouage de votre %@ devant votre %@.",
                san, PieceNaming.frenchKindLowercase(victim), PieceNaming.frenchKindLowercase(behind)
            )
        }

        guard mentionsCost, let materialLoss else { return head }
        return head + " " + LocalizationController.string("Vous perdez %lld points.", materialLoss)
    }

    /// Confort côté vue : suit le réglage de notation de l'utilisateur, comme
    /// ``SANFormatter/display(_:)``.
    @MainActor
    var sentence: String {
        sentence(notation: AppSettings.shared.pieceNotation)
    }
}

/// Fabrique l'explication d'un coup à partir de la réfutation du moteur.
///
/// Entièrement PUR : une position, une liste de coups en LAN, et rien d'autre.
/// Aucune requête moteur n'est faite ici — la variante est déjà payée par la
/// classification, qui interroge Stockfish sur la position d'après de toute
/// façon. L'explication est donc **gratuite en temps de calcul** ; c'est ce
/// qui la rend acceptable sur une revue de quarante coups.
enum MoveExplainer {

    /// Au-delà, la ligne ne raconte plus la faute mais la partie qui suit. Six
    /// coups couvrent largement une tactique.
    static let maxPlies = 12

    /// En deçà, la « perte » relève du bruit d'horizon : une variante coupée à
    /// une profondeur arbitraire peut montrer un pion d'écart qui n'existe pas.
    /// Même seuil que ``MoveClassifier/involvesSacrifice(move:boardAfterMove:)``,
    /// et pour la même raison.
    static let minimumMaterialLoss = 2

    struct Input {
        /// Position APRÈS le coup à expliquer — l'adversaire est au trait.
        var positionAfterMove: Position
        /// Variante principale du moteur à cette position, en LAN. Son premier
        /// élément est la réponse de l'adversaire : la punition.
        var refutationLANs: [String]
    }

    /// `nil` quand la ligne ne dit rien d'exploitable : ni mat, ni motif, ni
    /// perte matérielle nette. Mieux vaut se taire que meubler.
    static func explain(_ input: Input) -> MoveExplanation? {
        var board = Board(position: input.positionAfterMove)
        let mover = input.positionAfterMove.sideToMove.opposite
        let startingBalance = materialBalance(of: board.position, from: mover)

        var plies: [Move] = []
        /// Bilan matériel après chaque demi-coup, et si la position y est
        /// « calme » — c'est-à-dire si rien ne peut reprendre sur la case
        /// d'arrivée. Lire le matériel au milieu d'un échange donnerait un
        /// chiffre faux dans un sens ou dans l'autre.
        var balances: [(value: Int, isQuiet: Bool)] = []
        var boardAfterFirstPly: Board?
        var mateInMoves: Int?

        for lan in input.refutationLANs.prefix(maxPlies) {
            guard let move = apply(lan: lan, to: &board) else { break }
            plies.append(move)
            balances.append((
                value: materialBalance(of: board.position, from: mover),
                isQuiet: !canRecapture(on: move.end, in: board)
            ))
            if plies.count == 1 { boardAfterFirstPly = board }

            if case let .checkmate(matedColor) = board.state {
                // L'adversaire joue en premier dans la réfutation : il mate donc
                // aux demi-coups impairs (1, 3, 5…), soit (n + 1) / 2 coups.
                if matedColor == mover { mateInMoves = (plies.count + 1) / 2 }
                break
            }
        }

        guard let firstPly = plies.first, let boardAfterFirstPly else { return nil }

        let motif: TacticalMotif? = if let mateInMoves {
            .checkmate(
                inMoves: mateInMoves,
                isBackRank: TacticalMotifDetector.isBackRankMate(of: mover, board: board)
            )
        } else {
            TacticalMotifDetector.detect(punishing: firstPly, boardAfter: boardAfterFirstPly)
        }

        // Le verdict matériel se lit au dernier point CALME de la ligne. Sans
        // ça, une variante tronquée juste après une prise (limite de
        // profondeur ou fin de PV) annoncerait une perte de dame que la reprise
        // du demi-coup suivant aurait effacée.
        let settled = balances.last(where: \.isQuiet) ?? balances.last
        let loss = settled.map { startingBalance - $0.value }
        let materialLoss = (loss ?? 0) >= minimumMaterialLoss ? loss : nil

        guard motif != nil || materialLoss != nil else { return nil }
        return MoveExplanation(motif: motif, materialLoss: materialLoss, refutationSAN: firstPly.san)
    }

    // MARK: Utilitaires

    /// Bilan matériel du point de vue de `color` (positif = il a plus de bois).
    private static func materialBalance(of position: Position, from color: Piece.Color) -> Int {
        position.pieces.reduce(0) { total, piece in
            total + (piece.color == color ? pieceValue(piece.kind) : -pieceValue(piece.kind))
        }
    }

    /// Un camp peut-il reprendre sur `square` ? La case porte la pièce qui vient
    /// d'arriver, donc tout coup vers elle est une prise.
    private static func canRecapture(on square: Square, in board: Board) -> Bool {
        board.position.pieces
            .filter { $0.color == board.position.sideToMove }
            .contains { board.canMove(pieceAt: $0.square, to: square) }
    }

    /// Applique un coup en LAN (« e2e4 », « e7e8q »), promotion comprise, et
    /// rend le coup obtenu. Même logique que ``PuzzleSolutionTrimmer``.
    private static func apply(lan: String, to board: inout Board) -> Move? {
        guard lan.count >= 4 else { return nil }
        let start = Square(String(lan.prefix(2)))
        let end = Square(String(lan.dropFirst(2).prefix(2)))
        guard let move = board.move(pieceAt: start, to: end) else { return nil }
        if case .promotion = board.state {
            let kind: Piece.Kind = lan.count == 5
                ? (Piece.Kind(rawValue: String(lan.suffix(1)).uppercased()) ?? .queen)
                : .queen
            return board.completePromotion(of: move, to: kind)
        }
        return move
    }
}

File: android/app/src/main/java/com/chesslab/analysis/MoveExplanation.kt
package com.chesslab.analysis

import android.content.Context
import chesskit.Board
import chesskit.Move
import chesskit.Piece
import chesskit.Position
import chesskit.Square
import com.chesslab.R

/**
 * Pourquoi un coup était mauvais — en une phrase vraie. Pendant de
 * `MoveExplanation.swift`.
 *
 * Le bandeau coach sait dire « e5 — Erreur, −12 % » : il sait donc COMBIEN un
 * coup coûte, et ne sait pas dire CE QUI le punit. L'utilisateur apprend qu'il
 * a eu tort, pas ce qu'il n'a pas vu, et rejouera le même coup.
 *
 * La réponse est lue sur la RÉFUTATION DU MOTEUR — la variante qu'il enchaîne
 * après le coup joué. Aucun modèle de langage n'entre ici : tout est rejoué sur
 * un plateau, donc rien ne peut être inventé.
 */
data class MoveExplanation(
    /** Le motif qui punit, quand il est nommable. `null` est fréquent. */
    val motif: TacticalMotif?,
    /** Matériel net perdu sur la réfutation, POV du joueur qui vient de jouer. */
    val materialLoss: Int?,
    /** Le premier coup de la réfutation, en SAN. */
    val refutationSan: String,
) {
    /**
     * La phrase, dans la langue active. Construite à la LECTURE et non à
     * l'analyse : un changement de langue doit la retraduire, pas la laisser
     * figée dans celle qui avait cours au moment du calcul.
     */
    fun sentence(context: Context): String {
        val san = refutationSan
        val motif = motif
            ?: return materialLoss?.let { context.getString(R.string.explain_material, san, it) } ?: ""

        var mentionsCost = true
        val head = when (motif) {
            is TacticalMotif.Checkmate -> {
                // Le matériel ne veut plus rien dire quand la ligne mate.
                mentionsCost = false
                when {
                    motif.inMoves <= 1 && motif.isBackRank -> context.getString(R.string.explain_back_rank, san)
                    motif.inMoves <= 1 -> context.getString(R.string.explain_mate, san)
                    motif.isBackRank -> context.getString(R.string.explain_back_rank_in, san, motif.inMoves)
                    else -> context.getString(R.string.explain_mate_in, san, motif.inMoves)
                }
            }
            is TacticalMotif.HangingPiece -> {
                // Le coût est le nom de la pièce : le répéter en points serait du bruit.
                mentionsCost = false
                context.getString(R.string.explain_hanging, san, kindName(context, motif.kind))
            }
            is TacticalMotif.Fork -> {
                // Deux cibles nommées suffisent — une fourchette triple se
                // raconte aussi bien par ses deux plus grosses prises.
                val named = motif.targets.take(2).map { kindName(context, it) }
                if (named.size == 2) context.getString(R.string.explain_fork_two, san, named[0], named[1])
                else context.getString(R.string.explain_fork, san)
            }
            // La pièce qui démasque n'est PAS nommée : la flèche sur le plateau
            // montre déjà laquelle, et l'accord de genre coûterait un gabarit
            // par pièce pour un gain nul.
            is TacticalMotif.DiscoveredCheck -> context.getString(R.string.explain_discovered, san)
            // Tournure NOMINALE (« clouage de votre… ») et non « votre tour est
            // clouée » : le participe s'accorderait en genre avec la pièce, ce
            // qu'un gabarit à trous ne sait pas faire.
            is TacticalMotif.Pin -> context.getString(
                R.string.explain_pin, san, kindName(context, motif.victim), kindName(context, motif.behind)
            )
        }

        val loss = materialLoss
        if (!mentionsCost || loss == null) return head
        return head + " " + context.getString(R.string.explain_cost, loss)
    }

    /**
     * Le nom de la pièce EN MINUSCULES : la phrase la met au milieu (« votre
     * tour était en prise »), pas en tête. `lowercase()` sans locale, donc
     * invariant — les deux langues de l'app s'en accommodent.
     */
    private fun kindName(context: Context, kind: Piece.Kind): String = context.getString(
        when (kind) {
            Piece.Kind.king -> R.string.piece_king
            Piece.Kind.queen -> R.string.piece_queen
            Piece.Kind.rook -> R.string.piece_rook
            Piece.Kind.bishop -> R.string.piece_bishop
            Piece.Kind.knight -> R.string.piece_knight
            Piece.Kind.pawn -> R.string.piece_pawn
        }
    ).lowercase()
}

/**
 * Fabrique l'explication d'un coup à partir de la réfutation du moteur.
 *
 * Entièrement PUR : une position, une liste de coups en LAN, et rien d'autre.
 * Aucune requête moteur n'est faite ici — la variante est déjà payée par la
 * classification. L'explication est donc gratuite en temps de calcul, ce qui
 * la rend acceptable sur une revue de quarante coups.
 */
object MoveExplainer {

    /**
     * Au-delà, la ligne ne raconte plus la faute mais la partie qui suit.
     * Six coups couvrent largement une tactique.
     */
    const val MAX_PLIES = 12

    /**
     * En deçà, la « perte » relève du bruit d'horizon : une variante coupée à
     * une profondeur arbitraire peut montrer un pion d'écart qui n'existe pas.
     */
    const val MINIMUM_MATERIAL_LOSS = 2

    /**
     * @param positionAfterMove position APRÈS le coup à expliquer —
     *   l'adversaire est au trait.
     * @param refutationLans variante principale du moteur à cette position.
     *   Son premier élément est la réponse de l'adversaire : la punition.
     *
     * Rend `null` quand la ligne ne dit rien d'exploitable : ni mat, ni motif,
     * ni perte matérielle nette. Mieux vaut se taire que meubler.
     */
    fun explain(positionAfterMove: Position, refutationLans: List<String>): MoveExplanation? {
        val board = Board(positionAfterMove)
        val mover = positionAfterMove.sideToMove.opposite
        val startingBalance = materialBalance(board.position, mover)

        val plies = mutableListOf<Move>()
        /**
         * Bilan matériel après chaque demi-coup, et si la position y est
         * « calme » — c'est-à-dire si rien ne peut reprendre sur la case
         * d'arrivée. Lire le matériel au milieu d'un échange donnerait un
         * chiffre faux dans un sens ou dans l'autre.
         */
        val balances = mutableListOf<Pair<Int, Boolean>>()
        var boardAfterFirstPly: Board? = null
        var mateInMoves: Int? = null

        for (lan in refutationLans.take(MAX_PLIES)) {
            val move = apply(lan, board) ?: break
            plies += move
            balances += materialBalance(board.position, mover) to !canRecapture(move.end, board)
            if (plies.size == 1) boardAfterFirstPly = Board(board.position.copy())

            val state = board.state
            if (state is Board.State.Checkmate) {
                // L'adversaire joue en premier dans la réfutation : il mate donc
                // aux demi-coups impairs (1, 3, 5…), soit (n + 1) / 2 coups.
                if (state.color == mover) mateInMoves = (plies.size + 1) / 2
                break
            }
        }

        val firstPly = plies.firstOrNull() ?: return null
        val afterFirst = boardAfterFirstPly ?: return null

        val motif = mateInMoves?.let {
            TacticalMotif.Checkmate(it, TacticalMotifDetector.isBackRankMate(mover, board))
        } ?: TacticalMotifDetector.detect(firstPly, afterFirst)

        // Le verdict matériel se lit au dernier point CALME de la ligne. Sans
        // ça, une variante tronquée juste après une prise annoncerait une perte
        // de dame que la reprise du demi-coup suivant aurait effacée.
        val settled = balances.lastOrNull { it.second } ?: balances.lastOrNull()
        val loss = settled?.let { startingBalance - it.first }
        val materialLoss = if ((loss ?: 0) >= MINIMUM_MATERIAL_LOSS) loss else null

        if (motif == null && materialLoss == null) return null
        return MoveExplanation(motif, materialLoss, firstPly.san)
    }

    /** Bilan matériel du point de vue de [color] (positif = il a plus de bois). */
    private fun materialBalance(position: Position, color: Piece.Color): Int =
        position.pieces.sumOf { if (it.color == color) pieceValue(it.kind) else -pieceValue(it.kind) }

    /**
     * Un camp peut-il reprendre sur [square] ? La case porte la pièce qui vient
     * d'arriver, donc tout coup vers elle est une prise.
     */
    private fun canRecapture(square: Square, board: Board): Boolean =
        board.position.pieces
            .filter { it.color == board.position.sideToMove }
            .any { board.canMove(it.square, square) }

    /** Applique un coup en LAN (« e2e4 », « e7e8q »), promotion comprise. */
    private fun apply(lan: String, board: Board): Move? {
        if (lan.length < 4) return null
        val move = board.move(Square(lan.substring(0, 2)), Square(lan.substring(2, 4))) ?: return null
        if (board.state is Board.State.Promotion) {
            val kind = if (lan.length == 5) kindFor(lan.last()) else Piece.Kind.queen
            return board.completePromotion(move, kind)
        }
        return move
    }

    private fun kindFor(letter: Char): Piece.Kind = when (letter.lowercaseChar()) {
        'n' -> Piece.Kind.knight
        'b' -> Piece.Kind.bishop
        'r' -> Piece.Kind.rook
        else -> Piece.Kind.queen
    }
}
